File: Java/HTS221.java
// Distributed with a free-will license.
// Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
// HTS221
// This code is designed to work with the HTS221_I2CS I2C Mini Module available from ControlEverything.com.
// https://www.controleverything.com/content/Humidity?sku=HTS221_I2CS#tabs-0-product_tabset-2

import com.pi4j.io.i2c.I2CBus;
import com.pi4j.io.i2c.I2CDevice;
import com.pi4j.io.i2c.I2CFactory;
import java.io.IOException;

public class HTS221
{
	public static void main(String args[]) throws Exception
	{
		// Create I2CBus
		I2CBus bus = I2CFactory.getInstance(I2CBus.BUS_1);
		// Get I2C device, HTS221 I2C address is 0x5F(95)
		I2CDevice device = bus.getDevice(0x5F);

		// Select average configuration register
		// Temperature average samples = 16, humidity average samples = 32
		device.write(0x10, (byte)0x1B);
		// Select control register1
		// Power on, block data update, data rate o/p = 1 Hz
		device.write(0x20, (byte)0x85);
		Thread.sleep(500);

		// Read Calliberation values from the non-volatile memory of the device
		// Humidity calliberation values
		byte[] val = new byte[2];
		// Read 1 byte of data from address 0x30(48)
		val[0] = (byte)device.read(0x30);
		// Read 1 byte of data from address 0x31(49)
		val[1] = (byte)device.read(0x31);
		int H0 = (val[0] & 0xFF) / 2;
		int H1 = (val[1] & 0xFF) / 2;
		

		// Read 1 byte of data from address 0x36(54)
		val[0] = (byte)device.read(0x36);
		// Read 1 byte of data from address 0x37(55)
		val[1] = (byte)device.read(0x37);
		int H2 = ((val[1] & 0xFF) * 256) + (val[0] & 0xFF);

		// Read 1 byte of data from address 0x3A(58)
		val[0] = (byte)device.read(0x3A);
		// Read 1 byte of data from address 0x3B(59)
		val[1] = (byte)device.read(0x3B);
		int H3 = ((val[1] & 0xFF) * 256) + (val[0] & 0xFF);

		// Temperature calliberation values
		// Read 1 byte of data from address 0x32(50)
		int T0 = ((byte)device.read(0x32) & 0xFF);
		// Read 1 byte of data from address 0x33(51)
		int T1 = ((byte)device.read(0x33) & 0xFF);

		// Read 1 byte of data from address 0x35(53)
		int raw = ((byte)device.read(0x35) & 0x0F);
		
		// Convert the temperature calliberation values to 10-bits
		T0 = ((raw & 0x03) * 256) + T0;
		T1 = ((raw & 0x0C) * 64) + T1;

		// Read 1 byte of data from address 0x3C(60)
		val[0] = (byte)device.read(0x3C);
		// Read 1 byte of data from address 0x3D(61)
		val[1] = (byte)device.read(0x3D);
		int T2 = ((val[1] & 0xFF) * 256) + (val[0] & 0xFF);

		// Read 1 byte of data from address 0x3E(62)
		val[0] = (byte)device.read(0x3E);
		// Read 1 byte of data from address 0x3F(63)
		val[1] = (byte)device.read(0x3F);
		int T3 = ((val[1] & 0xFF) * 256) + (val[0] & 0xFF);
		
		// Read 4 bytes of data
		// hum msb, hum lsb, temp msb, temp lsb
		byte[] data = new byte[4];
		device.read(0x28 | 0x80, data, 0, 4);

		// Convert the data
		int hum = ((data[1] & 0xFF) * 256) + (data[0] & 0xFF); 
		int temp = ((data[3] & 0xFF) * 256) + (data[2] & 0xFF);
		if(temp > 32767)
		{
			temp -= 65536;
		}
		double humidity = ((1.0 * H1) - (1.0 * H0)) * (1.0 * hum - 1.0 * H2) / (1.0 * H3 - 1.0 * H2) + (1.0 * H0);
		double cTemp = ((T1 - T0) / 8.0) * (temp - T2) / (T3 - T2) + (T0 / 8.0);
		double fTemp = (cTemp * 1.8 ) + 32;

		// Output data to screen
		System.out.printf("Relative Humidity : %.2f %% RH %n", humidity);
		System.out.printf("Temperature in Celsius : %.2f C %n", cTemp);
		System.out.printf("Temperature in Farhenheit : %.2f F %n", fTemp);
	}
}

File: Python/HTS221.py
# Distributed with a free-will license.
# Use it any way you want, profit or free, provided it fits in the licenses of its associated works.
# HTS221
# This code is designed to work with the HTS221_I2CS I2C Mini Module available from ControlEverything.com.
# https://www.controleverything.com/content/Humidity?sku=HTS221_I2CS#tabs-0-product_tabset-2

import smbus
import time

# Get I2C bus
bus = smbus.SMBus(1)

# HTS221 address, 0x5F(95)
# Select average configuration register, 0x10(16)
#		0x1B(27)	Temperature average samples = 256, Humidity average samples = 512
bus.write_byte_data(0x5F, 0x10, 0x1B)
# HTS221 address, 0x5F(95)
# Select control register1, 0x20(32)
#		0x85(133)	Power ON, Continuous update, Data output rate = 1 Hz
bus.write_byte_data(0x5F, 0x20, 0x85)

time.sleep(0.5)

# HTS221 address, 0x5F(95)
# Read Calliberation values from non-volatile memory of the device
# Humidity calliberation values
# Read data back from 0x30(48), 1 byte
val = bus.read_byte_data(0x5F, 0x30)
H0 = val / 2

# Read data back from 0x31(49), 1 byte
val = bus.read_byte_data(0x5F, 0x31)
H1 = val /2

# Read data back from 0x36(54), 2 bytes
val0 = bus.read_byte_data(0x5F, 0x36)
val1 = bus.read_byte_data(0x5F, 0x37)
H2 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

# Read data back from 0x3A(58), 2 bytes
val0 = bus.read_byte_data(0x5F, 0x3A)
val1 = bus.read_byte_data(0x5F, 0x3B)
H3 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

# Temperature calliberation values
# Read data back from 0x32(50), 1 byte
T0 = bus.read_byte_data(0x5F, 0x32)
T0 = (T0 & 0xFF)

# Read data back from 0x32(51), 1 byte
T1 = bus.read_byte_data(0x5F, 0x33)
T1 = (T1 & 0xFF)

# Read data back from 0x35(53), 1 byte
raw = bus.read_byte_data(0x5F, 0x35)
raw = (raw & 0x0F)

# Convert the temperature calliberation values to 10-bits
T0 = ((raw & 0x03) * 256) + T0
T1 = ((raw & 0x0C) * 64) + T1

# Read data back from 0x3C(60), 2 bytes
val0 = bus.read_byte_data(0x5F, 0x3C)
val1 = bus.read_byte_data(0x5F, 0x3D)
T2 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

# Read data back from 0x3E(62), 2 bytes
val0 = bus.read_byte_data(0x5F, 0x3E)
val1 = bus.read_byte_data(0x5F, 0x3F)
T3 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

# Read data back from 0x28(40) with command register 0x80(128), 4 bytes
# humidity msb, humidity lsb, temp msb, temp lsb
data = bus.read_i2c_block_data(0x5F, 0x28 | 0x80, 4)

# Convert the data
humidity = (data[1] * 256) + data[0]
humidity = ((1.0 * H1) - (1.0 * H0)) * (1.0 * humidity - 1.0 * H2) / (1.0 * H3 - 1.0 * H2) + (1.0 * H0)
temp = (data[3] * 256) + data[2]
if temp > 32767 :
	temp -= 65536
cTemp = ((T1 - T0) / 8.0) * (temp - T2) / (T3 - T2) + (T0 / 8.0)
fTemp = (cTemp * 1.8 ) + 32

# Output data to screen
print "Relative Humidity : %.2f %%" %humidity
print "Temperature in Celsius : %.2f C" %cTemp
print "Temperature in Fahrenheit : %.2f F" %fTemp
